#### Instruction
- Use this template to develop your project. Do not change the steps. 
- For each step, you may add additional cells if needed.
- Describe the steps in the "Description:" field.

#### Group Information

Group No: Cancer_2

- Member 1: Foo Ling Zhen
- Member 2: Dean Lau Sheng Ting
- Member 3: Looi Wei En
- Member 4: Eason Peng 


#### Import libraries

In [51]:
%config Completer.use_jedi=False
import pandas as pd
import numpy as np 


from sklearn.model_selection import train_test_split
from sklearn.impute import SimpleImputer

#### Load the dataset

In [52]:
df= pd.read_csv('risk_factors.csv')
df

Age Number of sexual partners First sexual intercourse  \
0     18                       4.0                     15.0   
1     15                       1.0                     14.0   
2     34                       1.0                        ?   
3     52                       5.0                     16.0   
4     46                       3.0                     21.0   
..   ...                       ...                      ...   
853   34                       3.0                     18.0   
854   32                       2.0                     19.0   
855   25                       2.0                     17.0   
856   33                       2.0                     24.0   
857   29                       2.0                     20.0   

    Num of pregnancies Smokes Smokes (years) Smokes (packs/year)  \
0                  1.0    0.0            0.0                 0.0   
1                  1.0    0.0            0.0                 0.0   
2                  1.0    0.0            0.0                 0.0   
3                  4.0    1.0           37.0                37.0   
4                  4.0    0.0            0.0                 0.0   
..                 ...    ...            ...                 ...   
853                0.0    0.0            0.0                 0.0   
854                1.0    0.0            0.0                 0.0   
855                0.0    0.0            0.0                 0.0   
856                2.0    0.0            0.0                 0.0   
857                1.0    0.0            0.0                 0.0   

    Hormonal Contraceptives Hormonal Contraceptives (years)  IUD  ...  \
0                       0.0                             0.0  0.0  ...   
1                       0.0                             0.0  0.0  ...   
2                       0.0                             0.0  0.0  ...   
3                       1.0                             3.0  0.0  ...   
4                       1.0                            15.0  0.0  ...   
..                      ...                             ...  ...  ...   
853                     0.0                             0.0  0.0  ...   
854                     1.0                             8.0  0.0  ...   
855                     1.0                            0.08  0.0  ...   
856                     1.0                            0.08  0.0  ...   
857                     1.0                             0.5  0.0  ...   

    STDs: Time since first diagnosis STDs: Time since last diagnosis  \
0                                  ?                               ?   
1                                  ?                               ?   
2                                  ?                               ?   
3                                  ?                               ?   
4                                  ?                               ?   
..                               ...                             ...   
853                                ?                               ?   
854                                ?                               ?   
855                                ?                               ?   
856                                ?                               ?   
857                                ?                               ?   

    Dx:Cancer Dx:CIN Dx:HPV Dx Hinselmann Schiller Citology Biopsy  
0           0      0      0  0          0        0        0      0  
1           0      0      0  0          0        0        0      0  
2           0      0      0  0          0        0        0      0  
3           1      0      1  0          0        0        0      0  
4           0      0      0  0          0        0        0      0  
..        ...    ...    ... ..        ...      ...      ...    ...  
853         0      0      0  0          0        0        0      0  
854         0      0      0  0          0        0        0      0  
855         0      0      0  0          0        0        1      0  

In [53]:
df.shape

(858, 36)

In [54]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 858 entries, 0 to 857
Data columns (total 36 columns):
 #   Column                              Non-Null Count  Dtype 
---  ------                              --------------  ----- 
 0   Age                                 858 non-null    int64 
 1   Number of sexual partners           858 non-null    object
 2   First sexual intercourse            858 non-null    object
 3   Num of pregnancies                  858 non-null    object
 4   Smokes                              858 non-null    object
 5   Smokes (years)                      858 non-null    object
 6   Smokes (packs/year)                 858 non-null    object
 7   Hormonal Contraceptives             858 non-null    object
 8   Hormonal Contraceptives (years)     858 non-null    object
 9   IUD                                 858 non-null    object
 10  IUD (years)                         858 non-null    object
 11  STDs                                858 non-null    object

In [55]:
df.isna().sum()

Age                                   0
Number of sexual partners             0
First sexual intercourse              0
Num of pregnancies                    0
Smokes                                0
Smokes (years)                        0
Smokes (packs/year)                   0
Hormonal Contraceptives               0
Hormonal Contraceptives (years)       0
IUD                                   0
IUD (years)                           0
STDs                                  0
STDs (number)                         0
STDs:condylomatosis                   0
STDs:cervical condylomatosis          0
STDs:vaginal condylomatosis           0
STDs:vulvo-perineal condylomatosis    0
STDs:syphilis                         0
STDs:pelvic inflammatory disease      0
STDs:genital herpes                   0
STDs:molluscum contagiosum            0
STDs:AIDS                             0
STDs:HIV                              0
STDs:Hepatitis B                      0
STDs:HPV                              0


The dataframe has 858 entries and 36 columns, consisting of datatypes int64 and object. There no null values in the dataframe.

In [56]:
# Assigning features to X
X = df.drop('Biopsy', axis = 1)
X.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 858 entries, 0 to 857
Data columns (total 35 columns):
 #   Column                              Non-Null Count  Dtype 
---  ------                              --------------  ----- 
 0   Age                                 858 non-null    int64 
 1   Number of sexual partners           858 non-null    object
 2   First sexual intercourse            858 non-null    object
 3   Num of pregnancies                  858 non-null    object
 4   Smokes                              858 non-null    object
 5   Smokes (years)                      858 non-null    object
 6   Smokes (packs/year)                 858 non-null    object
 7   Hormonal Contraceptives             858 non-null    object
 8   Hormonal Contraceptives (years)     858 non-null    object
 9   IUD                                 858 non-null    object
 10  IUD (years)                         858 non-null    object
 11  STDs                                858 non-null    object

In [57]:
X.shape


(858, 35)

In [58]:
# Assigning column 'Biopsy' as the target, y
y = df['Biopsy']
y.info()

<class 'pandas.core.series.Series'>
RangeIndex: 858 entries, 0 to 857
Series name: Biopsy
Non-Null Count  Dtype
--------------  -----
858 non-null    int64
dtypes: int64(1)
memory usage: 6.8 KB


In [59]:
y.shape

(858,)

#### Split the dataset
Split the dataset into training, validation and test sets.

In [60]:
#split into 80:10:10

#splitting into training and test set
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.1,random_state=0)

#splitting training and validation set 
X_train, X_vald, y_train, y_vald = train_test_split(X, y, test_size = 0.1,random_state=0)


In [61]:
print(X_train.shape)
print(X_vald.shape)
print(X_test.shape)

(772, 35)
(86, 35)
(86, 35)


In [62]:
print(y_train.shape)
print(y_vald.shape)
print(y_test.shape)

(772,)
(86,)
(86,)


#### Data preprocessing
Perform data preprocessing such as normalization, standardization, label encoding etc.
______________________________________________________________________________________
Description:
In this data preprocessing operation, we perform the following in order:
1. Missing Value Detection and Replacement
2. Outlier Detection
3. Data Normalization
4. Data Standardization

##### Missing Value Detection and Replacement

In [63]:
# Show the datatypes of each column
X_train.dtypes

Age                                    int64
Number of sexual partners             object
First sexual intercourse              object
Num of pregnancies                    object
Smokes                                object
Smokes (years)                        object
Smokes (packs/year)                   object
Hormonal Contraceptives               object
Hormonal Contraceptives (years)       object
IUD                                   object
IUD (years)                           object
STDs                                  object
STDs (number)                         object
STDs:condylomatosis                   object
STDs:cervical condylomatosis          object
STDs:vaginal condylomatosis           object
STDs:vulvo-perineal condylomatosis    object
STDs:syphilis                         object
STDs:pelvic inflammatory disease      object
STDs:genital herpes                   object
STDs:molluscum contagiosum            object
STDs:AIDS                             object
STDs:HIV  

In [64]:
# Show the columns where datatype = 'object'
X_train.select_dtypes(include='object').head(10)

Number of sexual partners First sexual intercourse Num of pregnancies  \
301                         ?                     16.0                1.0   
789                       1.0                     18.0                1.0   
452                       1.0                     16.0                1.0   
344                       1.0                     17.0                1.0   
650                       2.0                     17.0                  ?   
524                       1.0                     17.0                3.0   
215                       1.0                     24.0                1.0   
620                       3.0                     18.0                2.0   
142                       3.0                     12.0                3.0   
141                       3.0                     18.0                1.0   

    Smokes Smokes (years) Smokes (packs/year) Hormonal Contraceptives  \
301    0.0            0.0                 0.0                     0.0   
789    0.0            0.0                 0.0                     0.0   
452    0.0            0.0                 0.0                       ?   
344    0.0            0.0                 0.0                     0.0   
650    1.0           10.0                 1.0                     1.0   
524    0.0            0.0                 0.0                     1.0   
215    0.0            0.0                 0.0                     1.0   
620    1.0            5.0        0.5132021277                     1.0   
142    1.0           16.0                12.0                       ?   
141    0.0            0.0                 0.0                       ?   

    Hormonal Contraceptives (years)  IUD IUD (years)  ... STDs:syphilis  \
301                             0.0  0.0         0.0  ...           0.0   
789                             0.0  0.0         0.0  ...           0.0   
452                               ?    ?           ?  ...             ?   
344                             0.0  0.0         0.0  ...           0.0   
650                             3.0  0.0         0.0  ...           0.0   
524                            0.16  1.0         5.0  ...           0.0   
215                            0.08  0.0         0.0  ...           0.0   
620                             1.0  0.0         0.0  ...           0.0   
142                               ?    ?           ?  ...             ?   
141                               ?    ?           ?  ...             ?   

    STDs:pelvic inflammatory disease STDs:genital herpes  \
301                              0.0                 0.0   
789                              0.0                 0.0   
452                                ?                   ?   
344                              0.0                 0.0   
650                              0.0                 0.0   
524                              0.0                 0.0   
215                              0.0                 0.0   
620                              0.0                 0.0   
142                                ?                   ?   
141                                ?                   ?   

    STDs:molluscum contagiosum STDs:AIDS STDs:HIV STDs:Hepatitis B STDs:HPV  \
301                        0.0       0.0      0.0              0.0      0.0   
789                        0.0       0.0      0.0              0.0      0.0   
452                          ?         ?        ?                ?        ?   
344                        0.0       0.0      0.0              0.0      0.0   
650                        0.0       0.0      0.0              0.0      0.0   
524                        0.0       0.0      0.0              0.0      0.0   
215                        0.0       0.0      0.0              0.0      0.0   
620                        0.0       0.0      0.0              0.0      0.0   
142                          ?         ?        ?                ?        ?   
141                          ?         ?        ?                ?        ?

We can observe that 26 columns have datatype 'object' although they consist mostly of floating point numbers. We can also see that some fields are filled with '?'. Thus, we can infer that the character '?' contributed to the heterogeneous data that resulted in the datatype 'object', and that '?' is used to represent missing values.

In [65]:
# Count the number of missing values in the columns
X_train[X_train == '?'].count()

Age                                     0
Number of sexual partners              26
First sexual intercourse                7
Num of pregnancies                     51
Smokes                                 12
Smokes (years)                         12
Smokes (packs/year)                    12
Hormonal Contraceptives                98
Hormonal Contraceptives (years)        98
IUD                                   104
IUD (years)                           104
STDs                                   94
STDs (number)                          94
STDs:condylomatosis                    94
STDs:cervical condylomatosis           94
STDs:vaginal condylomatosis            94
STDs:vulvo-perineal condylomatosis     94
STDs:syphilis                          94
STDs:pelvic inflammatory disease       94
STDs:genital herpes                    94
STDs:molluscum contagiosum             94
STDs:AIDS                              94
STDs:HIV                               94
STDs:Hepatitis B                  

In [66]:
# Replace '?' in the columns with nan
X_train.replace('?', np.nan, inplace=True)

# Create an instance of SimpleImputer with mean as the strategy
mean_imputer = SimpleImputer(strategy='mean')

# Fill in the nan values with the mean of their column values
X_train_imputed = pd.DataFrame(mean_imputer.fit_transform(X_train), columns=X_train.columns)


In [67]:
# Check for null values
X_train_imputed.isna().all()

Age                                   False
Number of sexual partners             False
First sexual intercourse              False
Num of pregnancies                    False
Smokes                                False
Smokes (years)                        False
Smokes (packs/year)                   False
Hormonal Contraceptives               False
Hormonal Contraceptives (years)       False
IUD                                   False
IUD (years)                           False
STDs                                  False
STDs (number)                         False
STDs:condylomatosis                   False
STDs:cervical condylomatosis          False
STDs:vaginal condylomatosis           False
STDs:vulvo-perineal condylomatosis    False
STDs:syphilis                         False
STDs:pelvic inflammatory disease      False
STDs:genital herpes                   False
STDs:molluscum contagiosum            False
STDs:AIDS                             False
STDs:HIV                        

In [68]:
# Check if any '?' remains in the columns
X_train_imputed[X_train_imputed=='?'].count()

Age                                   0
Number of sexual partners             0
First sexual intercourse              0
Num of pregnancies                    0
Smokes                                0
Smokes (years)                        0
Smokes (packs/year)                   0
Hormonal Contraceptives               0
Hormonal Contraceptives (years)       0
IUD                                   0
IUD (years)                           0
STDs                                  0
STDs (number)                         0
STDs:condylomatosis                   0
STDs:cervical condylomatosis          0
STDs:vaginal condylomatosis           0
STDs:vulvo-perineal condylomatosis    0
STDs:syphilis                         0
STDs:pelvic inflammatory disease      0
STDs:genital herpes                   0
STDs:molluscum contagiosum            0
STDs:AIDS                             0
STDs:HIV                              0
STDs:Hepatitis B                      0
STDs:HPV                              0


In [69]:
# Check if columns still consist of datatype = 'object'
X_train_imputed.dtypes

Age                                   float64
Number of sexual partners             float64
First sexual intercourse              float64
Num of pregnancies                    float64
Smokes                                float64
Smokes (years)                        float64
Smokes (packs/year)                   float64
Hormonal Contraceptives               float64
Hormonal Contraceptives (years)       float64
IUD                                   float64
IUD (years)                           float64
STDs                                  float64
STDs (number)                         float64
STDs:condylomatosis                   float64
STDs:cervical condylomatosis          float64
STDs:vaginal condylomatosis           float64
STDs:vulvo-perineal condylomatosis    float64
STDs:syphilis                         float64
STDs:pelvic inflammatory disease      float64
STDs:genital herpes                   float64
STDs:molluscum contagiosum            float64
STDs:AIDS                         

In [70]:
X_train_imputed


Age  Number of sexual partners  First sexual intercourse  \
0    20.0                   2.540214                      16.0   
1    20.0                   1.000000                      18.0   
2    17.0                   1.000000                      16.0   
3    18.0                   1.000000                      17.0   
4    26.0                   2.000000                      17.0   
..    ...                        ...                       ...   
767  24.0                   1.000000                      14.0   
768  27.0                   2.000000                      17.0   
769  24.0                   1.000000                      17.0   
770  20.0                   1.000000                      16.0   
771  46.0                   1.000000                      15.0   

     Num of pregnancies  Smokes  Smokes (years)  Smokes (packs/year)  \
0              1.000000     0.0             0.0                  0.0   
1              1.000000     0.0             0.0                  0.0   
2              1.000000     0.0             0.0                  0.0   
3              1.000000     0.0             0.0                  0.0   
4              2.245492     1.0            10.0                  1.0   
..                  ...     ...             ...                  ...   
767            0.000000     0.0             0.0                  0.0   
768            4.000000     0.0             0.0                  0.0   
769            3.000000     0.0             0.0                  0.0   
770            2.000000     0.0             0.0                  0.0   
771            6.000000     0.0             0.0                  0.0   

     Hormonal Contraceptives  Hormonal Contraceptives (years)       IUD  ...  \
0                   0.000000                         0.000000  0.000000  ...   
1                   0.000000                         0.000000  0.000000  ...   
2                   0.642433                         2.248609  0.106287  ...   
3                   0.000000                         0.000000  0.000000  ...   
4                   1.000000                         3.000000  0.000000  ...   
..                       ...                              ...       ...  ...   
767                 0.000000                         0.000000  0.000000  ...   
768                 1.000000                         2.000000  0.000000  ...   
769                 1.000000                         6.000000  0.000000  ...   
770                 0.000000                         0.000000  0.000000  ...   
771                 0.000000                         0.000000  0.000000  ...   

     STDs: Number of diagnosis  STDs: Time since first diagnosis  \
0                          0.0                          6.151515   
1                          1.0                          1.000000   
2                          0.0                          6.151515   
3                          0.0                          6.151515   
4                          0.0                          6.151515   
..                         ...                               ...   
767                        0.0                          6.151515   
768                        0.0                          6.151515   
769                        0.0                          6.151515   
770                        0.0                          6.151515   
771                        0.0                          6.151515   

     STDs: Time since last diagnosis  Dx:Cancer  Dx:CIN  Dx:HPV   Dx  \
0                            5.80303        0.0     0.0     0.0  0.0   
1                            1.00000        0.0     0.0     0.0  0.0   
2                            5.80303        0.0     0.0     0.0  0.0   
3                            5.80303        0.0     0.0     0.0  0.0   
4                            5.80303        0.0     0.0     0.0  0.0   
..                               ...        ...     ...     ...  ...   
767                          5.80303        0.0     0.0     0.0  0

In [71]:
X_train_imputed.describe(include='all')

Age  Number of sexual partners  First sexual intercourse  \
count  772.000000                 772.000000                772.000000   
mean    26.769430                   2.540214                 16.963399   
std      8.315672                   1.678928                  2.819925   
min     14.000000                   1.000000                 10.000000   
25%     20.000000                   2.000000                 15.000000   
50%     25.500000                   2.000000                 17.000000   
75%     32.000000                   3.000000                 18.000000   
max     79.000000                  28.000000                 32.000000   

       Num of pregnancies      Smokes  Smokes (years)  Smokes (packs/year)  \
count          772.000000  772.000000      772.000000           772.000000   
mean             2.245492    0.151316        1.253155             0.480533   
std              1.362623    0.355791        4.043737             2.303233   
min              0.000000    0.000000        0.000000             0.000000   
25%              1.000000    0.000000        0.000000             0.000000   
50%              2.000000    0.000000        0.000000             0.000000   
75%              3.000000    0.000000        0.000000             0.000000   
max             10.000000    1.000000       37.000000            37.000000   

       Hormonal Contraceptives  Hormonal Contraceptives (years)         IUD  \
count               772.000000                       772.000000  772.000000   
mean                  0.642433                         2.248609    0.106287   
std                   0.448121                         3.414818    0.286880   
min                   0.000000                         0.000000    0.000000   
25%                   0.000000                         0.000000    0.000000   
50%                   1.000000                         1.000000    0.000000   
75%                   1.000000                         2.248609    0.000000   
max                   1.000000                        22.000000    1.000000   

       ...  STDs: Number of diagnosis  STDs: Time since first diagnosis  \
count  ...                 772.000000                        772.000000   
mean   ...                   0.090674                          6.151515   
std    ...                   0.309078                          1.757875   
min    ...                   0.000000                          1.000000   
25%    ...                   0.000000                          6.151515   
50%    ...                   0.000000                          6.151515   
75%    ...                   0.000000                          6.151515   
max    ...                   3.000000                         22.000000   

       STDs: Time since last diagnosis   Dx:Cancer      Dx:CIN      Dx:HPV  \
count                       772.000000  772.000000  772.000000  772.000000   
mean                          5.803030    0.022021    0.010363    0.022021   
std                           1.715285    0.146846    0.101334    0.146846   
min                           1.000000    0.000000    0.000000    0.000000   
25%                           5.803030    0.000000    0.000000    0.000000   
50%                           5.803030    0.000000    0.000000    0.000000   
75%                           5.803030    0.000000    0.000000    0.000000   
max                          22.000000    1.000000    1.000000    1.000000   

               Dx  Hinselmann    Schiller    Citology  
count  772.000000  772.000000  772.000000  772.000000  
mean     0.028497    0.042746    0.088083    0.046632  
std      0.166497    0.202415    0.283599    0.210986  
min      0.000000    0.000000    0.000000    0.000000  
25%      0.000000    0.000000    0.000000    0.000000  
50%      0.000000    0.000000    0.000000    0.000000  
75%      0.000000    0.000000    0.000000    0.000000  
max      1.000000    1.000000    1.000000    1.000000  

[8 rows x 35 columns]

#### Feature Selection
Perform feature selection to select the relevant features.
______________________________________________________________________________________
Description:

#### Data modeling
Build the machine learning models. You must build atleast two (2) predictive models. One of the predictive models must be either Decision Tree or Support Vector Machine.
______________________________________________________________________________________
Description:

#### Evaluate the models
Perform a comparison between the predictive models. <br>
Report the accuracy, recall, precision and F1-score measures as well as the confusion matrix if it is a classification problem. <br>
Report the R2 score, mean squared error and mean absolute error if it is a regression problem.
______________________________________________________________________________________
Description: